eis_toolkit/transformations/minmax_scaling.py

```diff
@@ -0,0 +1,141 @@
+import numpy as np
+import pandas as pd
+import geopandas as gpd
+import rasterio
+from typing import Optional, Tuple, Union, List, Literal
+
+from eis_toolkit.exceptions import InvalidParameterValueException, InvalidInputDataException
+from eis_toolkit.transformations import utils
+from eis_toolkit.checks import parameter
+
+
```

Comment:
While trying to understand the code and come up with improvement suggestions, I ended up drafting an alternative implementation and used minmax as an example. Note that this is not tested code! Also, I made it mainly for you to get ideas, no need to take this suggestion super strictly.
```suggestion
@check_arguments
def minmax_scale(
    in_data: pd.DataFrame | gpd.GeoDataFrame | rasterio.io.DatasetReader,
    selection: Optional[List[int | str]] = None,
    new_ranges: List[Tuple[np.number, np.number]] = [(0, 1)]
) -> np.ndarray:
    
    if selection is not None and len(new_ranges) != 1 and len(selection) != len(new_ranges):
        raise InvalidParameterValueException("Specify scaling range for each selection or only one range")

    if isinstance(in_data, rasterio.DatasetReader):
        out_array = _minmax_scale_raster(in_data, selection, new_ranges)

    elif isinstance(in_data, pd.DataFrame | gpd.GeoDataFrame):
        out_array = _minmax_scale_vector(in_data, selection, new_ranges)
        
    return out_array


def _minmax_scale_raster(
    raster: rasterio.io.DatasetReader,
    selection: Optional[List[int]],
    new_ranges: List[Tuple[np.number, np.number]]
) -> np.ndarray:
    
    if selection is None:
        selection = list(range(1, raster.count + 1))
    else:
        check_bands_valid(raster, selection)

    out_array = raster.read()

    for band, new_range in expand_and_zip(selection, new_ranges):
        in_data = raster.read(band)
        nodata_values = raster.meta["nodata"]
        out_data = _minmax_scaling_core(in_data, nodata_values, new_range)
        out_array[band] = out_data

    return out_array


def _minmax_scale_vector(
    vector_data: pd.DataFrame | gpd.GeoDataFrame,
    selection: Optional[List[str]],
    new_ranges: List[Tuple[np.number, np.number]]
) -> np.ndarray:

    if selection is None:
        selection = vector_data.select_dtypes(include=[np.number]).columns.to_list()
        check_columns_valid(vector_data, selection)
    else:
        check_columns_valid(vector_data, selection)
        check_columns_numeric(vector_data, selection)

    out_data = vector_data.copy()

    for column, new_range in expand_and_zip(selection, new_ranges):
        series_as_array = out_data[column].to_numpy()
        nodata_values = "todo"
        out_data = _minmax_scaling_core(series_as_array, nodata_values, new_range)
        out_data[column] = out_data

    return out_data


@handle_nodata_as_nan
def _minmax_scaling_core(
    data: np.ndarray,
    new_range: Tuple[np.number, np.number],
) -> np.ndarray:
    
    min = np.nanmin(data)
    max = np.nanmax(data)
    scaled_min = new_range[0]
    scaled_max = new_range[1]
    
    scaler = (data - min) / (max - min)
    out_data = (scaler * (scaled_max - scaled_min)) + scaled_min
    
    return out_data
```